README.md
Minstall is a local module installer, intended to be used as postinstall-script.

## Example
Let's say you have the following modular app, and run `npm install` on it:
```
my-modular-app
├── modules
│   ├── database
│   │   ├── index.js
│   │   └── package.json [requires mongoose and lodash]
│   └── tasks
│       ├── index.js
│       └── package.json [requires lodash and database]
├── index.js
└── package.json [requires express, uses database and tasks]
```

#### The problems
- Your local modules (`database` and `tasks`) wouldn't work, because their dependencies are missing.
- To require your modules, you would need to either npm-link them, or use a `./modules/`-prefix

#### The solution
Minstall installs the necessary dependencies to the root-`node_modules`, and symlinks the modules there.
After running `npm install` with minstall as postinstall, the structure looks like this:
```
my-modular-app
├── modules
│   ├── database
│   │   ├── index.js
│   │   ├── node_modules
│   │   │   ├── lodash -> ../../../node_modules/lodash
│   │   │   └── mongoose -> ../../../node_modules/mongoose
│   │   └── package.json
│   └── tasks
│   │   ├── index.js
│   │   ├── node_modules
│   │   │   └── lodash -> ../../../node_modules/lodash
│   │   │   └── database -> ../../database
│   │   └── package.json
├── node_modules
│   ├── lodash
│   ├── minstall
│   └── mongoose
├── index.js
└── package.json
```
- All modules work, because their dependencies are present
  - conflicting dependencies end up in the associated modules, not in the root
- Modules can be required directly, because they are symlinked
  - ~~`require('./modules/database')`~~ -> `require('database')`
- The installation is faster and smaller, because dependencies are only installed once
  - dependencies that are already installed, are not re-downloaded
    (except for npm5 users, because of npm issue [#16853](https://github.com/npm/npm/issues/16853))


## Usage
- install with `npm install minstall --save`
- add it as postinstall-script to your package.json:
- `modules-folder` is optional, and defaults to `modules` if omitted
```JavaScript
"scripts": {
  "postinstall": "minstall <modules-folder>"
}
```

## Parameters
Minstall knows the following flags:
- `--no-link` prevents minstall from linking the local modules to the root-node_modules
- `--link-only` makes minstall go through the linking-process only, without installing anything
- `--cleanup` makes minstall remove all node_modules-folders before installing dependencies (this is forced for npm5)
- `--dependency-check-only` makes install print the dependency-check only, without touching any files or installing anything
- `--assume-local-modules-satisfy-non-semver-dependency-versions` (aka `--trust-local-modules`) makes minstall assume that a local module satisfies every requested version of that module that is not valid semver (like github-urls and tag-names)
- `--loglevel <loglevel>` sets the loglevel (`error`, `warn`, `info` `verbose`, `debug`, `silly`)
- `--no-hoist <dependency>`. makes minstall not hoist that dependency. `<dependency>` has the form name@versionRange, e.g. `--no-hoist aurelia-cli@^0.30.1`. If you omit the versionRange, no version of that dependency will be hoisted.
The name can be a glob expression (see [minimatch](https://www.npmjs.com/package/minimatch)), e.g. `--no-hoist aurelia-*`. This is useful for dependencies that don't play nice when hoisted/linked. This flag can be added multiple times.

## In collaboration with

![5Minds IT-Solutions](img/5minds_logo.png "5Minds IT-Solutions")

#### [5minds.de](https://5minds.de)

#### [github.com/5minds](https://github.com/5minds)


CHANGELOG.md
# 4.0.1
### Bugfixes
Fixed a bug where dependencies were not recognized as already satisfied

# 4.0.0
This is mainly a maintenance update. The API is unchanged, but i still mark it
as a major update, because support for node 4 (and probably node 6) is dropped
with this release.

### Improvements
Dependencies have been updated and minstall is now written in typescript.

# 3.3.0
### New Features
- new Flag `--no-hoist <dependency>`. Setting this makes minstall not hoist that
  dependency. `<dependency>` has the form name@versionRange, e.g.
  `--no-hoist aurelia-cli@^0.30.1`. If you omit the versionRange, no version of
  that dependency will be hoisted. The name can be a glob expression (see
  [minimatch](https://www.npmjs.com/package/minimatch)), e.g.
  `--no-hoist aurelia-*`. This is useful for dependencies that don't play nice
  when hoisted/linked. This flag can be added multiple times.

# 3.2.0
### New Features
- new Flag `--assume-local-modules-satisfy-non-semver-dependency-versions` (aka
  `--trust-local-modules`). Setting this makes minstall assume that a local
  module satisfies every requested version of that module that is not valid
  semver (like github-urls and tag-names)

### Bugfixes
- Fixes a bug where in rare cases minstall was wrongfully printing error-messages about installed packages not being found
- Added a workaround to work around https://github.com/snyamathi/semver-intersect/issues/7
- If an error occurs during installation, then minstall will now actually fail with exit-code 1

### Improvements
minstall will now exit without doing anything when it detects that you're using
npm 5.7.0. That version of npm has a very serious bug, see https://github.com/npm/npm/issues/19883

# 3.1.0
### New Features
- new `--link-only`-flag (makes minstall fix all linked dependencies (including links to local modules))

### Bugfixes
Minstall will now no longer try to hoist non-semver-dependencies, as minstall wouldn't be able to find them after the installation

### Improvements
The error-message when a package.json couldn't be parsed now includes the location of said package.json

# 3.0.4
### Bugfix
Minstall now no longer wrongully detects prerelease-versions as incompatible (Issue [#31](https://github.com/heikomat/minstall/issues/31))

# 3.0.3
### Improvements
Add support to use local modules within the parent-module itself (add
the local modules to the peerDependencies)

# 3.0.2
### Improvements
Update Readme-example for 3.0.x

# 3.0.1
### Improvements
fix a small typo in suboptimal-dependency-logs

# 3.0.0
### New Features
- new `--cleanup`-flag (makes minstall remove all node_modules-folders before installing dependencies)
- new `--dependency-check-only`-flag (makes install print the dependency-check only, without touching any files or installing anything)
- optimized detection of optimal dependency-installation-folder
- npm5-support through optimized dependency installation and forced `--cleanup`-flag
- parallel installation of conflicting dependencies

### Improvements
minstall 3 is way faster than minstall 2 because of better detection of optimized
dependency installation, parallel installation of conflicting dependencies and
npm5 support.

### Breaking changes
minstall now works completely different to version 2.0.0. Before you needed to
add minstall as postinstall to every level of local modules. Now only the
parent-project needs the minstall-postinstall.

**before:**

modules are found in parent-folders through node-module resolution
```
my-modular-app
├── modules
│   ├── database (@2.0.0) [requires abc@2.0.0 and tasks]
│   ├── tasks (@2.0.0) [requires abc@1.0.0, xyz, which in return requires database@1.0.0]
│   │   └── node_modules
│   │       └── abc@1.0.0
│   ├── test1(@2.0.0) [requires abc@1.0.0]
│   │   └── node_modules
│   │       └── abc@1.0.0
│   └── test2(@2.0.0) [requires abc@2.0.0]
├── node_modules
│   ├── abc@2.0.0
│   ├── minstall
│   ├── database ../modules/database
│   └── tasks -> ../modules/tasks
├── index.js
└── package.json [requires minstall, database@2.0.0 and tasks@2.0.0]
```

**now:**

the correct versions of the dependency are installed once and linked to the
destinations. in this example, abc@1.0.0 and abc@2.0.0 are only installed once,
while abc@1.0.0 would've been installed twice with minstall 2.0.2
```
my-modular-app
├── modules
│   ├── database (@2.0.0) [requires abc@2.0.0 and tasks]
│   │   └── node_modules
│   │       └── abc -> ../../../node_modules/abc
│   │       └── tasks -> ../../tasks
│   ├── tasks (@2.0.0) [requires abc@1.0.0, xyz, which in return requires database@1.0.0]
│   │   └── node_modules
│   │       ├── abc@1.0.0
│   │       ├── database@1.0.0
│   │       └── xyz
│   ├── test1 (@2.0.0) [requires abc@1.0.0]
│   │   └── node_modules
│   │       └── abc -> ../../tasks/node_modules/abc [<- NOT REINSTALLED, BUT LINKED!]
│   └── test2 (@2.0.0) [requires abc@2.0.0]
│       └── node_modules
│           └── abc -> ../../../node_modules/abc
├── node_modules
│   ├── abc@2.0.0
│   └── minstall
├── index.js
└── package.json [requires minstall, database@2.0.0 and tasks@2.0.0]
```

# 2.0.2
### Bugfixes
- Minstall now uses the folder name a local module should have according to the modules name
  to link it to node_modules, instead of its actual folder name ([#26](https://github.com/heikomat/minstall/pull/26))

# 2.0.1
### Improvements
- As of npm 5, npm automatically `--save`s dependencies installed with `npm install`.
  because minstall shouldn't touch the local modules package.json files, this change adds
  the `--no-save` flag the internaly used `npm install` command

# 2.0.0
### Bugfixes
- Fix linking of localy available packages, when they are also a sub-dependency (see below)

### Breaking changes
This breaking change applies **only** to the following scenario:
- linking is not disabled
- at some point somewhere you have a dependency on package xyz

src/minstall.ts
#!/usr/bin/env node

import * as os from 'os';
import * as path from 'path';
import * as readline from 'readline';
import * as semver from 'semver';
import * as winston from 'winston';
import {inspect} from 'util';
import {SPLAT} from 'triple-beam';

import {
  findOptimalDependencyTargetFolder,
  findRequestedDependencies,
  fixMissingDependenciesWithSymlinks,
  printNonOptimalDependencyInfos,
  printNonOptimalLocalModuleUsage,
  removeContradictingInstalledDependencies,
} from './dependency_handling';
import {
  DependencyInfo, DependencyRequests, DependencyTargetFolder, ModulesAndDependenciesInfo,
} from './interfaces';
import {ModuleInfo} from './module_info';
import {ModuleTools} from './moduletools';
import {SystemTools} from './systools';
import {UncriticalError} from './uncritical_error';

let logger: winston.Logger;
const setupLogger = (): void => {
  const logLevels = {
    critical: {level: 0, color: 'red'},
    error: {level: 1, color: 'magenta'},
    warn: {level: 2, color: 'yellow'},
    info: {level: 3, color: 'green'},
    verbose: {level: 4, color: 'gray'},
    debug: {level: 5, color: 'blue'},
    silly: {level: 6, color: 'cyan'},
  };

  const levels = {};
  const colors = {};
  for (const [name, level] of Object.entries(logLevels)) {
    levels[name] = level.level;
    colors[name] = level.color;
  }

  const isPrimitive = (value): boolean => {
    return value === null || (typeof value !== 'object' && typeof value !== 'function');
  };

  const formatWithInspect = (value): string => {
    const prefix = isPrimitive(value) ? '' : '\n';
    const shouldFormat = typeof value !== 'string';
    return prefix + (shouldFormat ? inspect(value, {depth: null, colors: true}) : value);
  };

  logger = winston.createLogger({
    levels: levels,
    transports: [
      new winston.transports.Console({
        stderrLevels: ['warn', 'error', 'critial'],
        format: winston.format.combine(
          winston.format.colorize(),
          winston.format.prettyPrint(),
          winston.format.simple(),
          winston.format.printf((info) => {
            const msg = formatWithInspect(info.message);
            const splatArgs = info[SPLAT] || [];
            const rest = splatArgs.map((data) => { return formatWithInspect(data); })
              .join(' ');

            return `${info.timestamp} - ${info.level}: ${msg} ${rest}`;
          }),
        ),
        handleExceptions: true,
      }),
    ],
  });
  winston.add(logger);
};

const cwd: string = process.cwd();

let commandConcatSymbol = ';';
let localPackage: ModuleInfo = null;

let installedAsDependency = false;
let projectFolderName: string = null;
let linkModules = true;
let npmVersion: string = null;
let cleanup = false;
let dependencyCheckOnly = false;
let linkOnly = false;
let isInProjectRoot = true;
let assumeLocalModulesSatisfyNonSemverDependencyVersions = false;
const noHoistList: Array<DependencyInfo> = [];

function logVerbose(): boolean {
  return ['verbose', 'debug', 'silly'].indexOf(logger.level) >= 0;
}

function logIfInRoot(message: string): void {
  if (message && message.length > 0 && (isInProjectRoot || logVerbose())) {
    // eslint-disable-next-line no-console
    console.log(message);
  }
}

function getLocalPackageInfo(): Promise<ModuleInfo> {
  return ModuleInfo.loadFromFolder(cwd, '');
}

async function _checkNpmVersion(): Promise<void> {
  npmVersion = await SystemTools.runCommand('npm --version', true);
  if (semver.satisfies(npmVersion, '5.7.0')) {
    // eslint-disable-next-line max-len
    logger.error('You\'re using npm 5.7.0. Do not use this version, it has a critical bug that is fixed in 5.7.1. See npm-issue #19883 for more info');
    process.exit(1);
  }

  // npm 5 workaround until npm-issue #16853 is fixed. replace the if-confition
  // and log when some npm version is confirmed working without that workaround.
  // even 5.7.1 is not working correctly without that workaround
  // if (semver.satisfies(npmVersion, '>=5.0.0 <5.7.0')) {
  const buggyNpmVersion = 5;
  if (semver.major(npmVersion) === buggyNpmVersion) {
    // logger.info('npm >=5.0.0 <5.7.0 detected. forcing --cleanup');
    logger.info('npm 5 detected. forcing --cleanup');
    cleanup = true;
  }
}

async function checkStartConditions(): Promise<void> {
  logger.debug('checking start conditions');

  await _checkNpmVersion();

  const pathParts: Array<string> = cwd.split(path.sep);
  let parentFolder: string;

  localPackage = await getLocalPackageInfo();
  if (!localPackage.isScoped) {
    const parentFolderIndexDifference = 2;
    parentFolder = pathParts[pathParts.length - parentFolderIndexDifference];
  } else {
    const scopedParentFolderIndexDifference = 3;
    parentFolder = pathParts[pathParts.length - scopedParentFolderIndexDifference];
  }

  if (parentFolder === 'node_modules') {
    logger.debug('project is in a node_modules folder. It\'s therefore installed as a dependency');
    installedAsDependency = true;
  }

  let folderName: string = await SystemTools.verifyFolderName(cwd, 'node_modules');
  if (folderName === null) {
    logger.debug('project folder has no node_modules-folder');

    if (!installedAsDependency) {
      logger.debug('project is not in a node_modules folder');
      throw new UncriticalError('minstall started from outside the project-root. aborting.');
    }
  } else {
    logger.debug('project folder has node_modules-folder');
  }

  folderName = '.';
  if (ModuleTools.modulesFolder !== '.') {
    folderName = await SystemTools.verifyFolderName(cwd, ModuleTools.modulesFolder);
  }

  if (folderName === null) {
    throw new UncriticalError(`${ModuleTools.modulesFolder} not found, thus minstall is done :)`);
  }

  if (installedAsDependency) {
    return;
  }

  isInProjectRoot = !await SystemTools.isSymlink(path.join(cwd, 'node_modules'));

  if (!localPackage.dependencies.minstall) {
    throw new UncriticalError('minstall started from outside the project-root. aborting.');
  }
}

function printInstallationStatus(startedInstallationCount: number, finishedInstallations: Array<number>): void {
  readline.clearLine(process.stdout, 0);
  readline.cursorTo(process.stdout, 0);
  const installationStatus: Array<string> = [];
  for (let index = 0; index < startedInstallationCount; index += 1) {
    if (finishedInstallations.indexOf(index) >= 0) {
      installationStatus.push(`${index + 1}: ✓`);
    } else {
      installationStatus.push(`${index + 1}:  `);
    }
  }
  process.stdout.write(installationStatus.join(' '));
}

async function installModuleDependencies(): Promise<void> {

  await removeContradictingInstalledDependencies();
  const targets: DependencyTargetFolder = await findOptimalDependencyTargetFolder(
    linkModules,
    assumeLocalModulesSatisfyNonSemverDependencyVersions,
    noHoistList,
  );

  // targets is an array where each entry has a location and a list of modules that should be installed
  const installPromises: Array<Promise<void>> = [];

  let startedInstallationCount = 0;
  const finishedInstallations: Array<number> = [];
  for (const [targetFolder, requestedDependencies] of Object.entries(targets)) {

    const shortTargetFolder = `.${targetFolder.substr(cwd.length)}`;
    const installationIndex: number = startedInstallationCount;
    startedInstallationCount += 1;

    logIfInRoot(`${installationIndex + 1}. installing ${requestedDependencies.length} dependencies to ${shortTargetFolder}`);
    installPromises.push(ModuleTools.installPackets(targetFolder, requestedDependencies)
      // eslint-disable-next-line no-loop-func
      .then(() => {
        finishedInstallations.push(installationIndex);
        printInstallationStatus(startedInstallationCount, finishedInstallations);
      }));
  }
  printInstallationStatus(Object.keys(targets).length, []);

  await Promise.all(installPromises);

  process.stdout.write('\n');

  // Now we're in a state where every dependency required by any local module
  // is installed at least somewhere. To make the modules find their dependencies
  // we now symlink them to the modules
  return fixMissingDependenciesWithSymlinks(linkModules, assumeLocalModulesSatisfyNonSemverDependencyVersions);

  // TODO: delete all unnecessary double-installs (when a and b both had a sub-dependency c that thus got installed twice)
  // this has low priority, as this is a rare edge-case with no real negative side-effects except a tiny bit bigger folder-size
}

async function runPostinstalls(): Promise<void> {
  const {modules: localModules} = await ModuleTools.getAllModulesAndInstalledDependenciesDeep();

  const postinstallPromises: Array<Promise<string>> = [];

  for (const module of localModules) {
    if (module.fullModulePath === cwd) {
      logger.debug('skipping the postinstall of the parent-module');
      continue;
    }

    if (!module.postinstallCommand) {
      logger.debug(`skipping the postinstall of ${module.name}. it has no postinstall script.`);
      continue;
    }

    logger.debug(`running postinstall of ${module.name}`);
    postinstallPromises.push(SystemTools.runCommand(`cd ${module.fullModulePath}${commandConcatSymbol} ${module.postinstallCommand}`));
  }

  await Promise.all(postinstallPromises);
}

async function deleteLinkedLocalModules(): Promise<void> {
  const moduleInfos: ModulesAndDependenciesInfo = await ModuleTools.getAllModulesAndInstalledDependenciesDeep();

  await Promise.all(moduleInfos.modules.map((moduleInfo: ModuleInfo) => {
    // local modules should be linked using the folder-names they should have,
    // no matter what folder-name they actually have, therefore don't use realFolderName here
    return SystemTools.delete(path.join(cwd, 'node_modules', moduleInfo.folderName));
  }));
}

function parseProcessArguments(): void {
  for (let i = 2; i < process.argv.length; i += 1) {
    if (process.argv[i].indexOf('--') !== 0) {
      ModuleTools.setModulesFolder(process.argv[i]);
    } else if (process.argv[i] === '--loglevel') {
      logger.level = process.argv[i + 1];
      i += 1;
    } else if (process.argv[i] === '--no-link') {
      linkModules = false;
    } else if (process.argv[i] === '--cleanup') {
      cleanup = true;
    } else if (process.argv[i] === '--dependency-check-only') {
      dependencyCheckOnly = true;
    } else if (process.argv[i] === '--assume-local-modules-satisfy-non-semver-dependency-versions' || process.argv[i] === '--trust-local-modules') {
      assumeLocalModulesSatisfyNonSemverDependencyVersions = true;
    } else if (process.argv[i] === '--link-only') {
      linkOnly = true;
    } else if (process.argv[i] === '--no-hoist') {
      const noHoistEntry: Array<string> = process.argv[i + 1].split('@');
      noHoistList.push({
        name: noHoistEntry[0],
        versionRange: noHoistEntry[1],
        identifier: `${noHoistEntry[0]}@${noHoistEntry[1]}`,
      });

      i += 1;
    }
  }
}

async function cleanupDependencies(): Promise<void> {
  const moduleInfos: ModulesAndDependenciesInfo = await ModuleTools.getAllModulesAndInstalledDependenciesDeep();

  await Promise.all(moduleInfos.modules.map((moduleInfo: ModuleInfo) => {
    // local modules should be linked using the folder-names they should have,
    // no matter what folder-name they actually have, therefore don't use realFolderName here
    return SystemTools.delete(path.join(moduleInfo.fullModulePath, 'node_modules'));
  }));
}

async function run(): Promise<void> {
  const startTime: number = Date.now();

  setupLogger();
  logger.level = 'info';
  parseProcessArguments();

  logger.silly('process arguments:', process.argv);
  logger.silly('os platfrom:', os.platform());
  logger.debug('loglevel:', logger.level);
  if (os.platform() === 'win32') {
    commandConcatSymbol = '&';
    ModuleTools.setNullTarget('NUL');
    ModuleTools.setCommandConcatSymbol(commandConcatSymbol);
  }

  const pathParts: Array<string> = cwd.split(path.sep);
  projectFolderName = pathParts[pathParts.length - 1];
  logger.debug('project folder name:', projectFolderName);

  if (dependencyCheckOnly) {
    const {modules: localModules} = await ModuleTools.getAllModulesAndInstalledDependenciesDeep();

    const requestedDependencies: DependencyRequests = findRequestedDependencies(localModules);
    printNonOptimalDependencyInfos(requestedDependencies);
    printNonOptimalLocalModuleUsage(localModules, requestedDependencies, assumeLocalModulesSatisfyNonSemverDependencyVersions);

    return;
  }

  if (linkOnly) {
    await fixMissingDependenciesWithSymlinks(linkModules, assumeLocalModulesSatisfyNonSemverDependencyVersions);
    return;
  }

  try {
    await checkStartConditions();
    if (linkModules) {
      await deleteLinkedLocalModules();
    }

    if (cleanup) {
      await cleanupDependencies();
    }

    await installModuleDependencies();
    await runPostinstalls();
    logIfInRoot(`\nminstall finished in ${SystemTools.getRuntime(startTime)} :)\n\n`);
  } catch (error) {
    if (error.constructor !== undefined && error.constructor.name === 'UncriticalError') {
      logIfInRoot(error.message);
    } else {
      throw error;
    }
  }
}

run();
